### PR TITLE
Provide common Java version detection

### PR DESCRIPTION
**Pull request by fvgh**
The approach of `GoogleJavaFormatter` to provide default versions depending of the JRE Java version has been extracted to a generic class. The code duplicated in `testlib` JRE has been removed.

Furthermore the approach has been extended:
- It allows to check whether a formatter version is supported by the current JRE. This approach allows to provide proposals about other formatter versions to the user. 
- It provides version proposals not requiring a JRE upgrade if possible.

For these extensions the amount of code has been increased. It needs to be discussed whether the benefit is worth the increase of code complexity.

### CODE REVIEW

File: lib/src/main/java/com/diffplug/spotless/Jvm.java
Comment (fvgh) [reply lib/src/main/java/com/diffplug/spotless/Jvm.java:110]: You are right, this is an unwanted bug. Logic should be:

1. `min-java-version`: `max-supported-formatter-version`
2. `min-java-version`: `max-supported-formatter-version`
3. `min-java-version`: `max-supported-formatter-version`
...

`min-java-version[n] > min-java-version[n+1]`
`max-supported-formatter-version[n] > max-supported-formatter-version[n+1]`

For n==0 the `max-supported-formatter-version` should be treated as the `max-known-formatter-version`,

File: lib/src/main/java/com/diffplug/spotless/java/GoogleJavaFormatStep.java
Comment (fvgh) [reply lib/src/main/java/com/diffplug/spotless/java/GoogleJavaFormatStep.java:102]: Yes, this is bug. See next comment.

**Comment by fvgh** [comment]
> release a new Spotless any time any one of our formatters releases a new version
I think it is beneficial for the users when default versions are increased frequently, since the user only needs to increase the Spotless version. But I think that everybody who automated the formatting with Spotless can wait for a few month for a new release. So just because someone updated a formatter version, I don't see the need that a new Spotless release follows a few days later.

This JVM check is just another assistance utility, since many projects still stuck with Java 8.

Not sure about your general remarks. Do you mean that instead of having a few utility functions, we should aim for something different?

**Comment by nedtwigg** [comment]
With your changes this whole thing looks great to me.

File: testlib/src/test/java/com/diffplug/spotless/JvmTest.java
Comment (fvgh) [inline testlib/src/test/java/com/diffplug/spotless/JvmTest.java:77]: Will be refactored once #900 is merged.

File: lib/src/main/java/com/diffplug/spotless/Jvm.java
Comment (fvgh) [inline lib/src/main/java/com/diffplug/spotless/Jvm.java:111]: @nedtwigg 
I moved unnecessary checks to an external method. My problem with `Jvm.Support` is the static configuration. It fits best with the current formatter step implementations, but it is not lazy. So I prefer to do sanity checks in the UTs.

**Comment by fvgh** [comment]
@nedtwigg Propose to do a "squash and merge". Didn't want to do a rebase and merged main instead to keep track of the conflicts.

File: testlib/src/test/java/com/diffplug/spotless/JvmTest.java
Comment (nedtwigg) [reply testlib/src/test/java/com/diffplug/spotless/JvmTest.java:77]: I think #900 is great, and I'm sorry that other PRs are causing merge conflicts for it, but I don't think we should hold other PRs up for it.

File: lib/src/main/java/com/diffplug/spotless/Jvm.java
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/Jvm.java:111]: Is your concern about laziness related to performance? I think navigating these tiny treemaps is going to be undetectable in terms of performance. If I understand your concern correctly (which maybe I don't), the tradeoff is:

- we could verify that the `Jvm.Support` metadata is consistent everytime they are used, which means we are adding microseconds of overhead to every single Spotless user forever
- we could add a unit test for every step that uses `Jvm.Support`
  - savings = microseconds of user time per build, rounds down to zero imo
  - costs = at least tens of minutes of contributor and reviewer time per PR

So not a big deal either way, but my preference is for less burden on the programmer if the only cost is microseconds of startup time.

**Review by nedtwigg — APPROVED** [review]
This is looking great to me :)

File: lib/src/main/java/com/diffplug/spotless/Jvm.java
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/Jvm.java:111]: Darn, my comment showed up twice, so I deleted one and they both got deleted :(

I think the sanity checks are fast enough that they cost ~0 to users, but adding a chore to the unit tests costs 10s of minutes per PR, so I lean very lightly towards making the verification automatic rather than adding them to a unit test checklist. I'm fine either way though, the core idea here is good.

**Comment by nedtwigg** [comment]
Btw, squash and merge sounds good, I'll let you push the button when you're ready. If there's more you want to put in this one then take your time :)

**Comment by fvgh** [comment]
@nedtwigg Tripped again over the circle-ci `test_nomaven...` problem, but saw that `com.diffplug.spotless:spotless-eclipse-groovy:4.2.0` was published.

However, I am afraid we made a mistake when closing #877.

I never published a `com.diffplug.spotless:spotless-eclipse-groovy:4.1.1`, nor did I do an update afterwards of the corresponding dependency locks in lib-extra. Hence we never provided a fixed version. Think we owe @k-brooks a big apology.

I will provide all formatter updates for Eclipse 4.20 at this weekend. But they will all require JVM 11+.

@k-brooks : Sorry, about the inconvenience. Is Eclipse 4.20 / JVM 11+ fine for you, or do you need JVM 8 support?

**Pull request merged**
